### PR TITLE
Support for controlling the device IME via nested scroll

### PR DESCRIPTION
**Pull request by chrisbanes**
ScreenFlow-optim

### CODE REVIEW

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt [hunk 1]
```diff
@@ -0,0 +1,201 @@
...
+            return available
+        }
+
+        if (available.y > 0) {
+            // If the user is scrolling down...
+
+            if (imeAnimController.isInsetAnimationInProgress()) {
+                // If we currently have control, we can update the IME insets using insetBy().
+                //
+                // The negation on the deltaY and the return value is necessary since nested
```
Comment (nickbutcher) [inline insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:104]: What negation?

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+    @delegate:RequiresApi(30)
+    private val imeAnimController by lazy(LazyThreadSafetyMode.NONE, ::SimpleImeAnimationController)
+
+    @get:RequiresApi(30)
+    private val imeVisible: Boolean
+        get() = view.rootWindowInsets.isVisible(WindowInsets.Type.ime())
+
+    override fun onPreScroll(available: Offset, source: NestedScrollSource): Offset {
+        if (Build.VERSION.SDK_INT < 30) {
+            // SimpleImeAnimationController only works on API 30+
```
Comment (nickbutcher) [inline insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:89]: Incorrect name?

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        return Offset.Zero
+    }
+
+    override fun onPostScroll(
+        consumed: Offset,
+        available: Offset,
+        source: NestedScrollSource
+    ): Offset {
+        if (Build.VERSION.SDK_INT < 30) {
+            // SimpleImeAnimationController only works on API 30+
```
Comment (nickbutcher) [inline insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:135]: Incorrect name?

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+        return Offset.Zero
+    }
+
+    override fun onPostFling(
+        consumed: Velocity,
+        available: Velocity,
+        onFinished: (Velocity) -> Unit
+    ) {
+        if (Build.VERSION.SDK_INT < 30) {
+            // SimpleImeAnimationController only works on API 30+
```
Comment (nickbutcher) [inline insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:172]: Incorrect name?

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+
+    @delegate:RequiresApi(30)
+    private val imeAnimController by lazy(LazyThreadSafetyMode.NONE, ::SimpleImeAnimationController)
+
+    @get:RequiresApi(30)
+    private val imeVisible: Boolean
+        get() = view.rootWindowInsets.isVisible(WindowInsets.Type.ime())
+
+    override fun onPreScroll(available: Offset, source: NestedScrollSource): Offset {
+        if (Build.VERSION.SDK_INT < 30) {
```
Comment (nickbutcher) [inline insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:88]: There are a number of API checks, could this class require API 30 to simplify it and move the check to when you create the ScrollConnection?

File: sample/src/main/java/dev/chrisbanes/accompanist/sample/insets/ImeAnimationSample.kt [hunk 6]
```diff
@@ -85,7 +87,9 @@ private fun Sample() {
             // FR for LazyColumn: https://issuetracker.google.com/173207790
             ScrollableColumn(
                 reverseScrollDirection = true,
-                modifier = Modifier.weight(1f)
+                modifier = Modifier
+                    .weight(1f)
+                    .nestedScroll(connection = rememberImeNestedScrollConnection())
```
Comment (nickbutcher) [inline sample/src/main/java/dev/chrisbanes/accompanist/sample/insets/ImeAnimationSample.kt:92]: 😍

File: insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt (hunk 5 above)
Comment (chrisbanes) [reply insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:88]: I thought about that but went this way for 2 reasons:

1. It's hard to do an `if (SDK_INT < 30)` when using fluent Modifier chains.
2. We _may_ eventually get a version of `WindowInsetsAnimationController` which works < 30 (doubtful but you never know)

Comment (nickbutcher) [reply insets/src/main/java/dev/chrisbanes/accompanist/insets/ImeNestedScrollConnection.kt:88]: Fair.

**Review by nickbutcher — APPROVED** [review]

**Pull request merged**